developers/development/guidelines.md

```diff
@@ -14,73 +14,191 @@ Exceptions can certainly be made, but they should be discussed and approved by a
-1. Logging levels should focus on the system itself and describe its state. As every bundle is only one out of many, logging should be done very scarce. It should be up to the user to increase the logging level for specific bundles, packages or classes if necessary. This means in detail:
- - Most logging should be done in `debug` level. `trace` can be used for even more details, where necessary.
- - Only few important things should be logged in `info` level, e.g. a newly started component or a user file that has been loaded.
- - `warn` logging should only be used to inform the user that something seems to be wrong in his overall setup, but the system can nonetheless function as normal, while possibly ignoring some faulty configuration/situation. It can also be used in situations, where a code section is reached, which is not expected by the implementation under normal circumstances (while being able to automatically recover from it).
- - `error` logging should only be used to inform the user that something is tremendously wrong in his setup, the system cannot function normally anymore, and there is a need for immediate action. It should also be used if some code fails irrecoverably and the user should report it as a severe bug.
-1. For bindings, you should NOT log errors, if e.g. connections are dropped - this is considered to be an external problem and from a system perspective to be a normal and expected situation. The correct way to inform users about such events is to update the Thing status accordingly. Note that all events (including Thing status events) are anyhow already logged.
-1. Likewise, bundles that accept external requests (such as servlets) must not log errors or warnings if incoming requests are incorrect. Instead, appropriate error responses should be returned to the caller.
+* Logging is not a replacement to use respective `update*` methods of the framework
 
-## Static Code Analysis
+```java
+void myFun() {
+  logger.debug("And now the thing goes online"); // DONT, DONT, really DONT do that
+  updateState(ThingState.ONLINE);
+}
+```
+
+Please keep the user informated via those `update*` methods if a connection drops etc and not abuse the logger.
+
+* `warn` logging should only be used to inform the user that something seems to be wrong in his overall setup, but the system can nonetheless function as normal, while possibly ignoring some faulty configuration/situation. It can also be used in situations, where a code section is reached, which is not expected by the implementation under normal circumstances (while being able to automatically recover from it).
+
+* `error` logging is not allowed in extensions and is purely reserved to the framework. Except if something is going really, really wrong in your extension and you are going to affect the framework stability.
```

Comment:
`error` logging is not allowed in extensions and is purely reserved to the framework. The only exception would be if something is going really, really wrong in your extension, and there is a possibility that the stability of the framework could be affected.